File: solar.py
import os,sys
userhome = os.getenv('HOME')
project_folder = 'gotm-dst'
base_folder=os.path.join(userhome,project_folder,'medsea_GOTM')
sys.path.append(os.path.join(userhome,project_folder,'python'))

from gotm_medsea import *
from datetime import datetime, timedelta 
from solar_utils import solposAM as solpos
from numpy import ones, array, cos, sin, pi, linspace, max, min, mean, argmax, argmin

# Global variable to be changed by user after loading.
year = 2014 # The year for which we are doing these computations, do not cross year boundary please.
timestep = 30 # in seconds

def tz(lon):
    if lon > 0:
        return int((lon+7.5)/15)
    else:
        return int((lon-7.5)/15)

def yrdays(year):
    return 366 if year % 4 == 0 else 365

def date2num(dt):
    from datetime import datetime
    ndays = (dt - datetime(dt.year,1,1)).days
    nsecs = (dt - datetime(dt.year,dt.month,dt.day,0,0,0)).seconds
    return ndays, nsecs

def UTC_to_local_nv(ndays,nsecs,lon):
    local_nsecs = nsecs + tz(lon)*3600
    local_ndays = ndays
    if local_nsecs > 86400:
        local_nsecs -= 86400
        local_ndays +=1
        
    elif local_nsecs < 0:
        local_nsecs += 86400
        local_ndays -= 1
    #if local_ndays < 0:
    #    print("WARNING: local date is in the previous year!")
    #if local_ndays > yrdays(year):
    #    print("WANRING: local date is in the next year!")
    
    return local_ndays, local_nsecs
    
def UTC_to_local(ndays,nsecs,lon):
    from numpy import vectorize
    vfunc = vectorize(UTC_to_local_nv)
    return vfunc(ndays,nsecs,lon)

def local_to_UTC_nv(ndays,nsecs,lon):
    UTC_nsecs = nsecs - tz(lon)*3600
    UTC_ndays = ndays
    if UTC_nsecs > 86400:
        UTC_nsecs -= 86400
        UTC_ndays +=1
    elif UTC_nsecs < 0:
        UTC_nsecs += 86400
        UTC_ndays -= 1
        #if UTC_ndays < 0:
        #   print("WARNING: UTC date is in the previous year!")
        #if UTC_ndays > yrdays(year):
        #   print("WARNING: UTC date is in the next year!")
    return UTC_ndays, UTC_nsecs 
        
def local_to_UTC(ndays,nsecs,lon):
    from numpy import vectorize
    vfunc = vectorize(local_to_UTC_nv)
    return vfunc(ndays,nsecs,lon)

## Low-accuracy General Solar Position Calculations by NOAA Global Monitoring Division [https://www.esrl.noaa.gov/gmd/grad/solcalc/solareqns.PDF. The value we need is coszen.

def gamma(ndays,nsecs):
    "Fractional solar year (in radians), which begins on Jan 1st noon. Assumes input in UTC."
    return 2*pi/yrdays(year)*(ndays+(nsecs/3600-12)/24)

def eqtime(t):
    "Equation of time (in minutes) as a function of fractional solar year (in radians)."
    # The NOAA version has 0.000075 which, according to Spencer via ... is incorrect.
    return 229.18*(0.0000075 + 0.001868*cos(t)   - 0.032077*sin(t) \
                             - 0.014615*cos(2*t) - 0.040849*sin(2*t))

def sundec(t):
    "Sun declination (in radians) as a function of fractional solar year (in radians)"
    return 0.006918 - 0.399912*cos(t) + 0.070257*sin(t) \
                    - 0.006758*cos(2*t) + 0.000907*sin(2*t) \
                    - 0.002697*cos(3*t) + 0.00148*sin(3*t)

def coszen_nv(ndays,nsecs,lat,lon):
    alat = lat/180*pi
    alon = lon/180*pi
    decl = sundec(gamma(*local_to_UTC(ndays,nsecs,lon)))
    #decl = sundec(gamma(ndays,nsecs))
    #lndays,lnsecs = UTC_to_local(ndays,nsecs,lon) # Which day it is should not matter.
    time_offset = eqtime(gamma(*local_to_UTC(ndays,nsecs,lon)))+4*lon-60*tz(lon)
    #time_offset = eqtime(gamma(ndays,nsecs))+4*lon-60*tz(lon)
    #tst = lnsecs/60.0 + time_offset
    #print(tz(lon),alon,time_offset,eqtime(gamma(*local_to_UTC(ndays,nsecs,lon))),eqtime(gamma(ndays,nsecs)))
    tst = nsecs/60.0 + time_offset
    ha = tst/4-180
    thsun = ha/180*pi
    return sin(alat)*sin(decl)+cos(alat)*cos(decl)*cos(thsun)        
        
def coszen(ndays,nsecs,lat,lon):
    "Cosine of the solar zenith angle. Inputs date and time should be local."
    from numpy import vectorize
    vfunc = vectorize(coszen_nv)
    return vfunc(ndays,nsecs,lat,lon)

def sp_coszen(ndays,nsecs,lat,lon):
    dt = datetime(year,1,1) + timedelta(days=ndays) + timedelta(seconds=nsecs)
    (angles, airmass) = solpos(location=[lat, lon, tz(lon)],
                               datetime=[dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second],
                               weather=[1015.62055, 40.0]) # Pressure and dry-bulb temp
    zenith, azimuth = angles # in degrees
    return cos(zenith/180*pi)

def coszen_argmax_error_seconds(ndays,lat,lon):
    ss = linspace(0,86400)
    sp_cc = array([sp_coszen(ndays,s,lat,lon) for s in ss])
    cc = coszen(ndays,ss,lat,lon)
    return ss[argmax(sp_cc)]-ss[argmax(cc)]

# Compute our swr using Rosati's formulas.

def swr_nv(ndays,nsecs,lat,lon):
    tau = 0.7  # atmospheric transmission coefficient (0.7 used in Rossatti)
    aozone = 0.09 # water vapour plus ozone absorption (0.09 used in Rossatti)
    solar = 1370  # solar constant
    cz = coszen_nv(ndays,nsecs,lat,lon)
    if cz <= 0:
        cz = 0.0
        qatten = 0.0
    else:
        qatten = tau**(1/cz)
    
    qzer  = cz*solar                       
    # Rosati (88) eq. (3.5-3.7)
    qdir  = qzer*qatten
    qdiff = ((1.-aozone)*qzer-qdir)*0.5
    qtot  = qdir+qdiff   
    return qtot    

def swr(ndays,nsecs,lat,lon):
    "Calculate short wave radiation for the given moment, time assumed to be local."
    from numpy import vectorize
#     assert (min(array(nsecs)) >= 0) and (max(array(nsecs))<=86400)
    vfunc = vectorize(swr_nv)
    return vfunc(ndays,nsecs,lat,lon)

def swr_3hourly_mean(ndays,nsecs,lat,lon,timestep):
    """Calculate short wave radiation, averaged for the subsequent 3-hourly period.
    Time period begins at the given moment (ndays, nsecs), assumed to be local.
    Average taken over equally spaced samples with duration of 
    'timestep' in seconds. 
    """
   
    # Number of samples
    ns = 3*3600/timestep
    assert ns-int(ns) == 0.0
    if ns-int(ns) == 0.0:
        ns = int(ns) 
    else: 
        raise Exception('The timestep given does not divide 3*3600 seconds.')

    # Find cumulative sum, then divide by # of samples to get mean.
    cumsum = 0 
    for i in range(ns):
        add_secs = timestep*i
        cumsum += swr_nv(ndays+int(nsecs/86400),
                         (nsecs+add_secs)%86400,
                         lat,lon)
    return cumsum/ns

def cloud_factor_calc(year,month,m,n,swr_ERA):
    from datetime import datetime
    from time import time
    from netCDF4 import Dataset
    from solar import swr_3hourly_mean
    import os
    GOTM_lat = linspace(30.75,45.75,21)
    GOTM_lon = linspace(-6.0,36.0,57)
    start = datetime(year,month,1)
    stop = datetime(year+1,1,1) if month == 12 else datetime(year,month+1,1)
    nrec = (stop-start).days*8
    ndays_start = (start-datetime(start.year,1,1)).days
    ndays_stop = (stop-datetime(start.year,1,1)).days
    tic = time()
    cloud_factor = ones((ndays_stop-ndays_start)*8) # Defaults to clear sky value.
    for ndays in range(ndays_start,ndays_stop):
        for i in range(8):
            nsecs = i*3*3600
            I_0_calc = swr_3hourly_mean(*UTC_to_local(ndays,nsecs,GOTM_lon[n]),GOTM_lat[m],GOTM_lon[n],timestep)
            ndays_of_month = ndays-ndays_start
            I_0_ERA = swr_ERA[ndays_of_month*8+i,m,n]
            # Instead of checking I_0_ERA, which can be zero for various reasons, check the clear sky value. 
            # If it's non-zero, compute the factor, otherwise, sun below horizon, so we just keep it as the defaul
            # value of 1 as initialized. Well, it should not matter.
            if I_0_calc > 1: 
                # If I_0_calc is really small but positive, we get into some trouble...
                cloud_factor[ndays_of_month*8+i] = I_0_ERA / I_0_calc
            if I_0_ERA < 0:
                # Use cloud_factor to zero out negative values of data.
                cloud_factor[ndays_of_month*8+i] = 0
    toc = time()
    # How long does it take for one grid point and one 3-hourly period?
    print('time elapsed for (m,n) = ({},{}):'.format(m,n), toc-tic)
    return cloud_factor

File: solar.f90
!EOC
!-----------------------------------------------------------------------
!BOP
!
! !IROUTINE: Calculate the short--wave radiation \label{sec:swr}
!
! !INTERFACE:
subroutine short_wave_radiation(jul,secs,alat,alon,swr)
  !
  ! !DESCRIPTION:
  !  This subroutine calculates the short--wave net radiation based on 
  !  latitude, longitude, time, fractional cloud cover and albedo.
  !  The albedo monthly values from \cite{Payne72} are given here
  !  as means of the values between 
  !  at 30$^{\circ}$ N and 40$^{\circ}$ N for the Atlantic Ocean 
  !  (hence the same latitudinal band of the Mediterranean Sea).
  !
  ! !USES:
  use time, only: calendar_date, julian_day, UTC_to_local, tz
  use airsea, only: albedo, solar

  ! WT 20171102, the following are the intended output variables after calling this subroutine.
  use airsea, only: qtot, cosr, coszen, sunbet, qdir_frac, qdiff_frac

  IMPLICIT NONE
  !
  ! !INPUT PARAMETERS:
  integer, intent(in)                 :: jul,secs
  double precision, intent(in)                :: alat,alon !WT assume radian
  !
  ! !OUTPUT PARAMETERS:
  double precision, optional, intent(out)     :: swr
  !   
  ! !REVISION HISTORY:
  !  Original author(s): Karsten Bolding
  !
  ! SH - 2003 - retain fractions of direct and diffuse radiation
  ! SH 08/2003 - keep a calculated 'clear sky' value of I_0 
  !
  !  See log for airsea module
  !
  !EOP
  !
  ! !LOCAL VARIABLES:
  !
  !   double precision                  :: solar=1350.

  ! WT local version of jul and secs, also lon in degrees.   
  !integer                           :: ljul,lsecs
  double precision                  :: lon,lat

  ! WT 20171102 Convenience parameters copied from airsea.f90
  double precision, parameter                :: pi=3.14159265358979323846
  double precision, parameter                :: deg2rad=pi/180.
  double precision, parameter                :: rad2deg=180./pi

  
  ! WT 20171102 Public module variables copied from airsea.f90
  !double precision, public	:: airt
  !double precision, public	:: solar
  !double precision, public	:: I_0_calc,coszen,cosr,qdir_frac,qdiff_frac
  !double precision, public      :: albedo,qtot,sunbet

  double precision                  :: eclips=23.439*deg2rad
  ! tau = 0.7 gives a better fit compared to 0.66 in medsea.
  double precision                  :: tau=0.7  !Arab=0.74,COARE=0.63,Sub=0.7
  double precision                  :: aozone=0.09

  double precision                  :: th0,th02,th03,thsun,solar_time,sundec !, coszen
  double precision                  :: gamma,tst,tst_offset,eqtime,ha,decl!, coszen, NOAA
  double precision                  :: zen,dzen  !,sunbet
  double precision                  :: qatten,qzer,qdir,qdiff,qshort
  double precision                  :: altitude !, qtot
  integer                   :: jab,count1,count2,k
  ! WT 20170315 Modifying new code by SP
  integer                   :: yyyy,mm,dd
  integer                           :: ljul, lsecs !local time
  integer                           :: jul0, jul1 !temp vars
  double precision                  :: yrdays,days,hours !WT renamed hour to hours

  double precision                  :: tjul
  double precision           ::alpha(1:480)

  !SP-cummulative days at each month
  !integer                   :: yday(12) = &
  !     (/ 0,31,59,90,120,151,181,212,243,273,304,334 /)

  !SP-values of albedo taken from Table 1 (Payne,72), with atmospheric transmittance of T=.70
  double precision                  :: alb1(46) = &
       (/.719,.656,.603,.480,.385,.300,.250,.193,.164,.145, &
       .131,.116,.103,.092,.084,.076,.071,.065,.061,.057, &
       .054,.051,.049,.039,.037,.035,.035,.035,.034,.033, &
       .033,.032,.032,.032,.029,.029,.029,.029,.028,.028, &
       .028,.028,.027,.027,.028,.028/)

  !SP-corresponding values of sun altitude
  double precision                  :: alt(46) = &
       (/0.0,2.,4.,6.,8.,10.,12.,14.,16.,18.,20.,22.,24.,26.,28., &
       30.,32.,34.,36.,38.,40.,42.,44.,46.,48.,50.,52.,54.,56., &
       58.,60.,62.,64.,66.,68.,70.,72.,74.,76.,78.,80.,82.,84.,86., &
       88.,90./)
  !
  !
  !-----------------------------------------------------------------------
  !BOC

  !Convert back to degrees
  lon = alon / deg2rad
  lat = alat / deg2rad

!!! WT 20170316 The sundec formula below is defined from the UTC time zone.
  !   The following effort to to find local values of date and times in hours,
  !   number of seconds since midnight etc... are no longer necessary.
  !
  !
  ! 1. Find the local calendar date: yyyy-mm-dd.
  !print*, "UTM ", jul,secs
  !call UTC_to_local(jul,secs,lon,ljul,lsecs)
  !print *, "jul,secs,lon,ljul,lsecs", jul,secs,lon,ljul,lsecs
  !print*, "local ", ljul,lsecs
  !call calendar_date(ljul,yyyy,mm,dd)
  !call calendar_date(jul,yyyy,mm,dd)
  !
  ! 2. Find the julian day of the final day of last year, save to ljul0.
  !call julian_day(yyyy-1,12,31,ljul0)
  !
  ! 3. Now get the day number of the local day of year.
  !days=float(yday(mm))+float(dd) ! SP's version.
  !days = float(ljul-ljul0) ! int to float, for later formulas
  !print *, "yyyy,mm,dd,days,ljul,ljul0", yyyy,mm,dd,days,ljul,ljul0
  !print *, "days, ljul-ljul0", days, ljul-ljul0 ! Should be the same.

  !kbk   if (mod(yy,4) .eq. 0 ! leap year I forgot
  !yrdays=365. ! GOTM's version.

  ! 4. Find the julian day number of the final day of this year, save to ljul1.
  !call julian_day(yyyy,12,31,ljul1)
  !
  ! 5. Find the number of days in this year.
  !yrdays = float(ljul1-ljul0) ! int to float, for later formulas
  !print *, "yrdays",yrdays
  !
  ! !!! WT End of the previous calculations using local timezone.

!!! Calculation of true solar time to find clear sky value of solar SWR, I_0_calc
  !
  ! Sources:
  ! https://www.esrl.noaa.gov/gmd/grad/solcalc/solareqns.PDF
  ! https://arxiv.org/pdf/1102.3825.pdf

  ! 1. Find calendar date from julian day number in UTC: yyyy-mm-dd, and yrdays of that year
  call calendar_date(jul,yyyy,mm,dd)
  if (mod(yyyy,4) .eq. 0) then
     yrdays = 366
  else
     yrdays = 365
  end if

  ! 2. Find number of days since yyyy-01-01, and the fractional hour since midnight.
  call julian_day(yyyy,1,1,jul0) ! jul0 = julian day number of the first day of year.
  days = jul - jul0 ! = day_of_year-1 = 0 for the first day yyyy:01-01
  hours = 1.*secs/3600. ! hours should be UTC decimal time

  ! 3. The fractional solar year (\gamma) in solareqns.pdf, which begins noon on civil New Year Day. 
  gamma = (2.*pi/yrdays)*(days+((hours-12)/24))  

  ! 4. Finding sun declination, the angle between the equator and sun ray.
  ! The Spencer formula (Spencer, 1971):
  decl = 0.006918 - 0.399912*cos(gamma)    + 0.070257*sin(gamma)        &
       - 0.006758*cos(2.*gamma) + 0.000907*sin(2.*gamma)     &
       - 0.002697*cos(3.*gamma) + 0.001480*sin(3.*gamma)  ! in radians

  !th0 = (2.*pi/yrdays)*(days-1+((hour-12)/24))  ! hour should be UTC decimal time
  !! An alternative (crude)
  !!th0 = 2.*pi*days/yrdays
  !
  !th02 = 2.*th0
  !th03 = 3.*th0
  !sundec = 0.006918 - 0.399912*cos(th0) + 0.070257*sin(th0)         &
  !     - 0.006758*cos(th02) + 0.000907*sin(th02)                 &
  !     - 0.002697*cos(th03) + 0.001480*sin(th03)  ! in radians
  !An alternative
  !sundec = 23.45*pi / 180.*sin(2.*pi*(284. + days) / 365.)

  ! 5. Equation of time, i.e. difference between apparent solar time (by sundials),
  ! and mean solar time (by civil calendar & clock).
  ! * the coefficient 0.0000075 is correct: http://www.mail-archive.com/sundial@uni-koeln.de/msg01050.html
  ! * the erroneous coefficient 0.000075 is reproduced in multiple documents, including the NOAA pdf quoted above.
  !eqtime = 229.18*(0.0000075+0.001868*cos(th0)-0.032077*sin(th0)-0.014615*cos(th02)-0.040849*sin(th02))  ! in minutes
  eqtime = 229.18*(0.0000075 + 0.001868*cos(gamma)    - 0.032077*sin(gamma) &
       - 0.014615*cos(2.*gamma) - 0.040849*sin(2.*gamma))  ! in minutes
  ! An alternative
  !IF ((days.GE.1).AND.(days.LE.106)) THEN
  !    eqtime = -14.2*sin(pi*(days+7.)/111.)
  !ELSE IF ((days.GE.107).AND.(days.LE.166)) THEN
  !    eqtime = 4.0*sin(pi*(days-106.)/59.)
  !ELSE IF ((days.GE.167).AND.(days.LE.246)) THEN
  !    eqtime = -6.5*sin(pi*(days-166.)/80.)
  !ELSE IF ((days.GE.247).AND.(days.LE.365)) THEN
  !    eqtime = 16.4*sin(pi*(days-247.)/113.)
  !END IF


  ! WT Be careful, does hour mean the hour of day, or the fractional number of hours since midnight?
  !solar_time = hour + (eqtime/60) + ((30.-lon)/15) ! in hours
  !solar_time = hour + (eqtime/60) - lon/15 ! in hours (UTC time), lon=degrees
  !solar_time = hour*60 +eqtime + 4*lon - (lsecs-secs)/60 ! The value should not exceed 1440.

  ! 5. Find true solar time in local time at a position with longitude = `lon`.
  ! Along prime meridian, with theoretical timezone tz = 0, the only offset is due to eqtime. Elsewhere, it's the difference between the longitudinal minute and the timezone offset, which are also zeros at multiples of 15 degress if tz is the theoretical timezone only depending on longitude.
  tst_offset = eqtime + 4.0*lon - 60.0*tz(lon)
  call UTC_to_local(jul,secs,lon,ljul,lsecs) ! Get the local time no. of seconds since midnight.
  tst = lsecs/60.0 + tst_offset ! True solar time, in minutes.

  !WT No need for getting the true tst, ha differ by 360 degrees
  ! after adjusting tst by 1440 minutes. It will not change value of cos(ha)
  ! if (tst > 24*60) then
  !    ! ignore the change in calendar date which does not affect the following
  !    tst = tst - 24*60 
  ! else if (tst < 0 ) then
  !    tst = tst + 24*60
  ! end if

  !print *, "lsecs,tst_offset,tst",lsecs,tst_offset,tst
  !print *,"tst,solar_time",tst,solar_time

  !  sun hour angle :
  !   thsun = (hour-12.)*15.*deg2rad + alon
  !thsun = (hour-12.)*15.*deg2rad
  !thsun = (12.-hour)*pi/12.
  !thsun = (12.-solar_time)*pi/12.
  !thsun = pi*((solar_time/12)-1)  ! radians
  !thsun = pi*((solar_time/(4*180))-1)  ! radians

  ! 6. Find the solar hour angle.
  ha = (tst/4-180)*deg2rad ! radians
  !print *,"thsun,ha,thsun-ha",thsun,ha,thsun-ha
  !PRINT*, thsun

  ! 7. Cosine of the solar zenith angle
  !(Rosati(88) eq. 3.4 :
  !coszen =sin(alat)*sin(sundec)+cos(alat)*cos(sundec)*cos(thsun)
  coszen =sin(alat)*sin(decl)+cos(alat)*cos(decl)*cos(ha) ! ha, decl not needed anymore from now on
  !WT 20170316 Consider moving the above to time.f90 or a new solar.f90 for later reuse (e.g. assimilation
  ! at sunrise or sunset, compare the SST turnaround times with solar sunrise and sunset...)
  !We could also use NREL's solpos C code also if we know how to recompile everything in separate dynamic linkable
  !library files. 


  if (coszen .le. 0.0) then
     coszen = 0.0           !SP-this is when sun is below horizon
     qatten = 0.0           !i.e between dusk and dawn
  else
     qatten = tau**(1./coszen)
  end if

  qzer  = coszen * solar                       
  qdir  = qzer * qatten                        !Rosati(88) eq. (3.5)
  qdiff = ((1.-aozone)*qzer - qdir) * 0.5      !Rosati(88) eq. (3.6)
  qtot  =  qdir + qdiff                        !Rosati(88) eq. (3.7)

!!!!! SH - 2003 - retain fractions of direct and diffuse radiation
  ! ensure fractions remain finite
  if (qtot .gt. 0) then 
     qdir_frac = qdir/qtot
     qdiff_frac = qdiff/qtot
  else
     qdir_frac = 0.
     qdiff_frac = 0.
  end if

!!!!!

  tjul = (days-81.)/yrdays*2.*pi

  !  sin of the solar noon altitude in radians (Rosati, 88) eq. 3.9:
  sunbet=sin(alat)*sin(eclips*sin(tjul))+cos(alat)*cos(eclips*sin(tjul))
  !  solar noon altitude in degrees :
  sunbet = asin(sunbet)*rad2deg

  !  calculates the albedo as a function of sun altitude :
  !  (after Payne jas 1972)
  !  solar zenith angle in degrees :
  zen=(180./pi)*acos(coszen)
  !  sun altitude :
  altitude=90.-zen

  jab=0.5*altitude + 1.

  !linear interpolation
  albedo=alb1(jab)+.5*(alb1(jab+1)-alb1(jab))*(altitude-alt(jab))      

  !SH  calculate cosine of angle of direct refracted entrant radiation 
  cosr = cos(asin((3./4.)*sin(acos(coszen))))

  !  radiation as from Reed(1977), Simpson and Paulson(1979)
  !  calculates SHORT WAVE FLUX ( watt/m*m )
  !  Rosati,Miyakoda 1988 ; eq. 3.8
  !  clouds from COADS perpetual data set


  !170301 Reed formula used here.
  !   if(cloud .lt.0.3) then
  !   if(cloud.eq.0.0) then
  !      qshort  = qtot*(1.-albedo)        !SP albedo factor needed here
  !   else !170301 consider removing this case... over the top?
  !      qshort  = qtot*(1.-.62*cloud + .0019*sunbet)*(1.-albedo)
  !   end if
  ! print *, 'qtot, qshort, albedo', qtot, qshort, albedo
  qshort=qtot*(1.-albedo)

  if (present(swr)) then
     swr = qshort
  else
     !     I_0_calc = qshort
     !     I_0_cs   = qtot*(1.-albedo)
     write(0,*) 'WARNING: short_wave_radiation() called without specifying an output variable name'
  end if

  return
end subroutine short_wave_radiation
     !EOC
